pallets/liquidity-pools-gateway/queue/src/lib.rs

```diff
@@ -0,0 +1,276 @@
+	pub type MessageQueue<T: Config> = StorageMap<_, Blake2_128Concat, T::MessageNonce, T::Message>;
+
+	/// Storage for messages that failed during processing.
+	#[pallet::storage]
+	#[pallet::getter(fn failed_message_queue)]
+	pub type FailedMessageQueue<T: Config> =
+		StorageMap<_, Blake2_128Concat, T::MessageNonce, (T::Message, DispatchError)>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub (super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// A message was submitted.
+		MessageSubmitted {
+			nonce: T::MessageNonce,
+			message: T::Message,
+		},
+
+		/// Message execution success.
+		MessageExecutionSuccess {
+			nonce: T::MessageNonce,
+			message: T::Message,
+		},
+
+		/// Message execution failure.
+		MessageExecutionFailure {
+			nonce: T::MessageNonce,
+			message: T::Message,
+			error: DispatchError,
+		},
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Message not found.
+		MessageNotFound,
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_idle(_now: BlockNumberFor<T>, max_weight: Weight) -> Weight {
+			Self::service_message_queue(max_weight)
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Convenience method for manually processing a message.
+		///
+		/// If the execution fails, the message gets moved to the
+		/// `FailedMessageQueue` storage.
+		///
+		/// NOTE - this extrinsic does not error out during message processing
+		/// to ensure that any storage changes (i.e. to the message queues)
+		/// are not reverted.
+		#[pallet::weight(T::WeightInfo::process_message())]
+		#[pallet::call_index(0)]
+		pub fn process_message(
+			origin: OriginFor<T>,
+			nonce: T::MessageNonce,
+		) -> DispatchResultWithPostInfo {
+			ensure_signed(origin)?;
+
+			let message = MessageQueue::<T>::take(nonce).ok_or(Error::<T>::MessageNotFound)?;
+
+			match Self::process_message_and_deposit_event(nonce, message.clone()) {
+				Ok(p) => Ok(p),
+				Err(e) => {
+					FailedMessageQueue::<T>::insert(nonce, (message, e.error));
+
+					Ok(e.post_info)
+				}
+			}
+		}
+
+		/// Convenience method for manually processing a failed message.
+		///
+		/// If the execution is successful, the message gets removed from the
+		/// `FailedMessageQueue` storage.
+		///
+		/// NOTE - this extrinsic does not error out during message processing
+		/// to ensure that any storage changes (i.e. to the message queues)
+		/// are not reverted.
+		#[pallet::weight(T::WeightInfo::process_failed_message())]
+		#[pallet::call_index(1)]
+		pub fn process_failed_message(
+			origin: OriginFor<T>,
+			nonce: T::MessageNonce,
+		) -> DispatchResultWithPostInfo {
+			ensure_signed(origin)?;
+
+			let (message, _) =
+				FailedMessageQueue::<T>::get(nonce).ok_or(Error::<T>::MessageNotFound)?;
+
+			match Self::process_message_and_deposit_event(nonce, message.clone()) {
+				Ok(p) => {
+					FailedMessageQueue::<T>::remove(nonce);
+
+					Ok(p)
+				}
+				Err(e) => Ok(e.post_info),
+			}
+		}
+	}
```

Comment:
So, for example, if we return an error instead of `Ok(e.post_info)` what's the difference in the storage? If I read this part well, nothing has changed. Then, why not just error out and warn the user that it was wrong instead of telling them that "extrinsics was succesful" ?

Reply:
In this case, for this extrinsic, we won't have the `MessageExecutionFailure` event.

Reply:
Ok, so you want to record that event in those cases

Reply:
As a retry history

Reply:
Yes, it's basically the exact same behavior we have in the gateway, and given that these are more like convenience methods, I was thinking that we can live with them.

Reply:
> these are more like convenience methods

Makes more sense having that in mind!